In [40]:
import os
import pandas as pd

# Mostrar desde dónde se ejecuta el notebook
print("Directorio actual:")
print(os.getcwd())

Directorio actual:
c:\Users\nayen\EP2-pipeline-end-to-end-dashboard\models\notebooks


In [41]:
# Ruta al dataset procesado
DATA_PATH = "../../data/processed/dataset_movilidad_rm.csv"

df = pd.read_csv(DATA_PATH)

print("Dataset cargado correctamente")
print(f"Filas: {df.shape[0]}")
print(f"Columnas: {df.shape[1]}")

Dataset cargado correctamente
Filas: 5000
Columnas: 16


In [42]:
df.head()

,id,fecha_hora,comuna_id,comuna,ingreso_promedio_hogar,paradero_id,codigo_paradero,latitud,longitud,recorrido_id,recorrido,empresa_operadora,capacidad_pasajeros,tiempo_promedio_viaje,flujo_pasajeros,velocidad_promedio
0,1,2025-01-01 00:00:00,5,La Florida,1100000.0,5,LF105,-33.5200,-70.5800,1,210,Metbus,120,34.45,330,37.25
1,2,2025-01-01 01:00:00,7,Providencia,2200000.0,7,PR107,-33.4300,-70.6100,3,506,Redbus,110,32.06,605,34.96
2,3,2025-01-01 02:00:00,5,La Florida,1100000.0,5,LF105,-33.5200,-70.5800,6,Metro L1,Metro,1000,20.34,1143,50.43
3,4,2025-01-01 03:00:00,7,Providencia,2200000.0,7,PR107,-33.4300,-70.6100,7,Metro L4,Metro,1000,19.56,1302,49.79
4,5,2025-01-01 04:00:00,1,Puente Alto,850000.0,1,PA101,-33.6111,-70.5755,8,Metro L5,Metro,1000,16.01,265,53.94


In [43]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      5000 non-null   int64  
 1   fecha_hora              5000 non-null   str    
 2   comuna_id               5000 non-null   int64  
 3   comuna                  5000 non-null   str    
 4   ingreso_promedio_hogar  5000 non-null   float64
 5   paradero_id             5000 non-null   int64  
 6   codigo_paradero         5000 non-null   str    
 7   latitud                 5000 non-null   float64
 8   longitud                5000 non-null   float64
 9   recorrido_id            5000 non-null   int64  
 10  recorrido               5000 non-null   str    
 11  empresa_operadora       5000 non-null   str    
 12  capacidad_pasajeros     5000 non-null   int64  
 13  tiempo_promedio_viaje   5000 non-null   float64
 14  flujo_pasajeros         5000 non-null   int64  
 15

In [44]:
df.describe(include="all")

,id,fecha_hora,comuna_id,comuna,ingreso_promedio_hogar,paradero_id,codigo_paradero,latitud,longitud,recorrido_id,recorrido,empresa_operadora,capacidad_pasajeros,tiempo_promedio_viaje,flujo_pasajeros,velocidad_promedio
count,5000.000000,5000,5000.000000,5000,5.000000e+03,5000.000000,5000,5000.000000,5000.000000,5000.000000,5000,5000,5000.000000,5000.000000,5000.000000,5000.000000
unique,NaN,5000,NaN,10,NaN,NaN,10,NaN,NaN,NaN,10,5,NaN,NaN,NaN,NaN
top,NaN,2025-01-01 00:00:00,NaN,Maipu,NaN,NaN,MA102,NaN,NaN,NaN,Metro L5,Metro,NaN,NaN,NaN,NaN
freq,NaN,1,NaN,922,NaN,NaN,922,NaN,NaN,NaN,551,2520,NaN,NaN,NaN,NaN
mean,2500.500000,NaN,3.844200,NaN,1.244200e+06,3.844200,NaN,-33.514740,-70.653912,5.516000,NaN,NaN,560.071000,32.392034,1215.051000,40.102890
std,1443.520003,NaN,2.325641,NaN,7.520794e+05,2.325641,NaN,0.071921,0.077577,2.872245,NaN,NaN,443.538324,15.987660,780.045581,11.304509
min,1.000000,NaN,1.000000,NaN,8.000000e+05,1.000000,NaN,-33.611100,-70.760000,1.000000,NaN,NaN,100.000000,13.830000,200.000000,10.000000
25%,1250.750000,NaN,2.000000,NaN,8.500000e+05,2.000000,NaN,-33.590000,-70.740000,3.000000,NaN,NaN,115.000000,19.077500,641.000000,31.895000
50%,2500.500000,NaN,3.000000,NaN,9.500000e+05,3.000000,NaN,-33.510000,-70.666700,6.000000,NaN,NaN,1000.000000,27.220000,999.000000,40.405000
75%,3750.250000,NaN,5.000000,NaN,1.100000e+06,5.000000,NaN,-33.450000,-70.575500,8.000000,NaN,NaN,1000.000000,41.350000,1641.250000,50.972500


Definir variable objetivo

In [45]:
# Variable objetivo (lo que queremos predecir)
TARGET = "tiempo_promedio_viaje"

# Variables predictoras
X = df.drop(columns=[TARGET])

# Variable objetivo
y = df[TARGET]

print("Variable objetivo:", TARGET)
print("Dimensión de X:", X.shape)
print("Dimensión de y:", y.shape)

Variable objetivo: tiempo_promedio_viaje
Dimensión de X: (5000, 15)
Dimensión de y: (5000,)


Eliminar columnas

In [46]:
X = X.drop(columns=[
    "id",
    "codigo_paradero"
])

X.head()

,fecha_hora,comuna_id,comuna,ingreso_promedio_hogar,paradero_id,latitud,longitud,recorrido_id,recorrido,empresa_operadora,capacidad_pasajeros,flujo_pasajeros,velocidad_promedio
0,2025-01-01 00:00:00,5,La Florida,1100000.0,5,-33.5200,-70.5800,1,210,Metbus,120,330,37.25
1,2025-01-01 01:00:00,7,Providencia,2200000.0,7,-33.4300,-70.6100,3,506,Redbus,110,605,34.96
2,2025-01-01 02:00:00,5,La Florida,1100000.0,5,-33.5200,-70.5800,6,Metro L1,Metro,1000,1143,50.43
3,2025-01-01 03:00:00,7,Providencia,2200000.0,7,-33.4300,-70.6100,7,Metro L4,Metro,1000,1302,49.79
4,2025-01-01 04:00:00,1,Puente Alto,850000.0,1,-33.6111,-70.5755,8,Metro L5,Metro,1000,265,53.94


Transformar la fecha

In [47]:
X["fecha_hora"] = pd.to_datetime(X["fecha_hora"])

X["hora"] = X["fecha_hora"].dt.hour
X["dia"] = X["fecha_hora"].dt.day
X["mes"] = X["fecha_hora"].dt.month

X = X.drop(columns=["fecha_hora"])

X.head()

,comuna_id,comuna,ingreso_promedio_hogar,paradero_id,latitud,longitud,recorrido_id,recorrido,empresa_operadora,capacidad_pasajeros,flujo_pasajeros,velocidad_promedio,hora,dia,mes
0,5,La Florida,1100000.0,5,-33.5200,-70.5800,1,210,Metbus,120,330,37.25,0,1,1
1,7,Providencia,2200000.0,7,-33.4300,-70.6100,3,506,Redbus,110,605,34.96,1,1,1
2,5,La Florida,1100000.0,5,-33.5200,-70.5800,6,Metro L1,Metro,1000,1143,50.43,2,1,1
3,7,Providencia,2200000.0,7,-33.4300,-70.6100,7,Metro L4,Metro,1000,1302,49.79,3,1,1
4,1,Puente Alto,850000.0,1,-33.6111,-70.5755,8,Metro L5,Metro,1000,265,53.94,4,1,1


Identificar variables

In [48]:
# Variables numéricas
numeric_features = X.select_dtypes(include=["int64", "float64"]).columns.tolist()

# Variables categóricas
categorical_features = X.select_dtypes(include=["object"]).columns.tolist()

print("Variables numéricas:")
print(numeric_features)

print("\nVariables categóricas:")
print(categorical_features)

Variables numéricas:
['comuna_id', 'ingreso_promedio_hogar', 'paradero_id', 'latitud', 'longitud', 'recorrido_id', 'capacidad_pasajeros', 'flujo_pasajeros', 'velocidad_promedio']

Variables categóricas:
['comuna', 'recorrido', 'empresa_operadora']


C:\Users\nayen\AppData\Local\Temp\ipykernel_38812\444075714.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_features = X.select_dtypes(include=["object"]).columns.tolist()


Comprobar el resultado

In [49]:
print(f"Variables numéricas: {len(numeric_features)}")
print(f"Variables categóricas: {len(categorical_features)}")

X.info()

Variables numéricas: 9
Variables categóricas: 3
<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   comuna_id               5000 non-null   int64  
 1   comuna                  5000 non-null   str    
 2   ingreso_promedio_hogar  5000 non-null   float64
 3   paradero_id             5000 non-null   int64  
 4   latitud                 5000 non-null   float64
 5   longitud                5000 non-null   float64
 6   recorrido_id            5000 non-null   int64  
 7   recorrido               5000 non-null   str    
 8   empresa_operadora       5000 non-null   str    
 9   capacidad_pasajeros     5000 non-null   int64  
 10  flujo_pasajeros         5000 non-null   int64  
 11  velocidad_promedio      5000 non-null   float64
 12  hora                    5000 non-null   int32  
 13  dia                     5000 non-null   int32  
 14  mes

Pipeline de Scikit-Learn

In [50]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler

Crear el preprocesador

In [51]:
# Transformaciones para variables numéricas
numeric_transformer = StandardScaler()

# Transformaciones para variables categóricas
categorical_transformer = OneHotEncoder(handle_unknown="ignore")

# Preprocesador
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features),
    ]
)

print("Preprocesador creado correctamente.")

Preprocesador creado correctamente.


Se utilizó un ColumnTransformer para aplicar un escalado (StandardScaler) a las variables numéricas y una codificación (OneHotEncoder) a las variables categóricas. Esto permite que el Pipeline procese automáticamente distintos tipos de datos sin necesidad de transformaciones manuales y evita el riesgo de Data Leakage.

Separar datos de entrenamiento y prueba

In [52]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(f"Entrenamiento: {X_train.shape}")
print(f"Prueba: {X_test.shape}")

Entrenamiento: (4000, 15)
Prueba: (1000, 15)


primer modelo (Decision Tree)

In [53]:
from sklearn.tree import DecisionTreeRegressor

# Crear Pipeline completo
dt_pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", DecisionTreeRegressor(random_state=42))
])

# Entrenar
dt_pipeline.fit(X_train, y_train)

print("Modelo entrenado correctamente.")

Modelo entrenado correctamente.


Realizar predicciones

In [54]:
y_pred = dt_pipeline.predict(X_test)

print("Predicciones realizadas correctamente.")

Predicciones realizadas correctamente.


Evaluar el modelo

In [55]:
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred) ** 0.5
r2 = r2_score(y_test, y_pred)

print(f"MAE : {mae:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"R²  : {r2:.4f}")

MAE : 2.29
RMSE: 3.03
R²  : 0.9624


Un R² de 0.9624 significa que el modelo explica aproximadamente el 96,24 % de la variabilidad del tiempo promedio de viaje.

Modelo Random Forest

In [56]:
from sklearn.ensemble import RandomForestRegressor

rf_pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", RandomForestRegressor(
        random_state=42,
        n_estimators=100
    ))
])

rf_pipeline.fit(X_train, y_train)

print("Random Forest entrenado correctamente.")

Random Forest entrenado correctamente.


In [57]:
y_pred_rf = rf_pipeline.predict(X_test)

In [58]:
mae_rf = mean_absolute_error(y_test, y_pred_rf)
rmse_rf = mean_squared_error(y_test, y_pred_rf) ** 0.5
r2_rf = r2_score(y_test, y_pred_rf)

print(f"MAE : {mae_rf:.2f}")
print(f"RMSE: {rmse_rf:.2f}")
print(f"R²  : {r2_rf:.4f}")

MAE : 1.78
RMSE: 2.30
R²  : 0.9782


Se incorporó Random Forest porque reduce el sobreajuste típico de un único árbol de decisión. En lugar de depender de un solo árbol, combina múltiples árboles entrenados sobre diferentes subconjuntos de datos, logrando predicciones más robustas y estables

El Random Forest supera al Decision Tree en todas las métricas:

El MAE disminuyó de 2.29 a 1.78, por lo que el error promedio es menor.
El RMSE también se redujo, indicando que los errores grandes son menos frecuentes.
El R² aumentó de 0.9624 a 0.9782, lo que significa que el modelo explica aproximadamente el 97,82 % de la variabilidad del tiempo promedio de viaje.

Tercer modelo Gradient Boosting

In [59]:
from sklearn.ensemble import GradientBoostingRegressor

gb_pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", GradientBoostingRegressor(
        random_state=42
    ))
])

gb_pipeline.fit(X_train, y_train)

print("Gradient Boosting entrenado correctamente.")

Gradient Boosting entrenado correctamente.


In [60]:
y_pred_gb = gb_pipeline.predict(X_test)

In [61]:
mae_gb = mean_absolute_error(y_test, y_pred_gb)
rmse_gb = mean_squared_error(y_test, y_pred_gb) ** 0.5
r2_gb = r2_score(y_test, y_pred_gb)

print(f"MAE : {mae_gb:.2f}")
print(f"RMSE: {rmse_gb:.2f}")
print(f"R²  : {r2_gb:.4f}")

MAE : 1.67
RMSE: 2.12
R²  : 0.9816


Comparación de modelos

In [62]:
import pandas as pd

resultados = pd.DataFrame({
    "Modelo": [
        "Decision Tree",
        "Random Forest",
        "Gradient Boosting"
    ],
    "MAE": [
        mae,
        mae_rf,
        mae_gb
    ],
    "RMSE": [
        rmse,
        rmse_rf,
        rmse_gb
    ],
    "R²": [
        r2,
        r2_rf,
        r2_gb
    ]
})

resultados = resultados.sort_values(by="R²", ascending=False)

resultados

,Modelo,MAE,RMSE,R²
2,Gradient Boosting,1.674233,2.117794,0.981593
1,Random Forest,1.775417,2.304965,0.978195
0,Decision Tree,2.289025,3.028708,0.962352


Conclusión del experimento

Se entrenaron tres algoritmos de regresión utilizando el mismo conjunto de entrenamiento y el mismo proceso de preprocesamiento mediante un Pipeline de Scikit-Learn. El modelo Gradient Boosting Regressor obtuvo el mejor desempeño, alcanzando un R² de 0.9816, además del menor error absoluto medio (MAE = 1.67) y el menor error cuadrático medio (RMSE = 2.12). Por este motivo, fue seleccionado para la etapa de optimización mediante GridSearchCV e integración con la API REST.

Optimizar el mejor modelo con GridSearchCV

In [63]:

from sklearn.model_selection import GridSearchCV

Pipeline para optimizar

In [64]:
gb_pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", GradientBoostingRegressor(random_state=42))
])

Definir los hiperparámetros

In [65]:
param_grid = {
    "model__n_estimators": [100, 200],
    "model__learning_rate": [0.05, 0.1],
    "model__max_depth": [3, 5]
}

Ejecutar GridSearchCV

In [66]:
grid_search = GridSearchCV(
    estimator=gb_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="r2",
    n_jobs=-1
)

grid_search.fit(X_train, y_train)

print("Optimización finalizada.")

Optimización finalizada.


Mejores parámetros

In [67]:
print("Mejores parámetros encontrados:")
print(grid_search.best_params_)

print("\nMejor R² obtenido en validación cruzada:")
print(f"{grid_search.best_score_:.4f}")

Mejores parámetros encontrados:
{'model__learning_rate': 0.05, 'model__max_depth': 3, 'model__n_estimators': 100}

Mejor R² obtenido en validación cruzada:
0.9795


Evaluar el mejor modelo

In [68]:
best_model = grid_search.best_estimator_

y_pred_best = best_model.predict(X_test)

mae_best = mean_absolute_error(y_test, y_pred_best)
rmse_best = mean_squared_error(y_test, y_pred_best) ** 0.5
r2_best = r2_score(y_test, y_pred_best)

print(f"MAE : {mae_best:.2f}")
print(f"RMSE: {rmse_best:.2f}")
print(f"R²  : {r2_best:.4f}")

MAE : 1.68
RMSE: 2.12
R²  : 0.9816


Se aplicó GridSearchCV con validación cruzada de 5 particiones para optimizar el modelo Gradient Boosting. Los resultados mostraron que el modelo base ya presentaba un desempeño cercano al óptimo, obteniendo métricas prácticamente iguales después de la optimización. Esto indica que el modelo inicial estaba adecuadamente configurado y no requería ajustes significativos.

In [69]:
import os
import joblib

# Crear carpeta si no existe
os.makedirs("../artifacts", exist_ok=True)

# Guardar el mejor modelo
joblib.dump(best_model, "../artifacts/best_model.pkl")

print("Modelo guardado correctamente.")

Modelo guardado correctamente.


In [70]:
import joblib

modelo = joblib.load("../artifacts/best_model.pkl")

print(type(modelo))

<class 'sklearn.pipeline.Pipeline'>


In [71]:
import json

metrics = {
    "modelo": "Gradient Boosting",
    "mae": round(mae_best, 2),
    "rmse": round(rmse_best, 2),
    "r2": round(r2_best, 4),
    "best_params": grid_search.best_params_
}

with open("../artifacts/metrics.json", "w", encoding="utf-8") as f:
    json.dump(metrics, f, indent=4, ensure_ascii=False)

print("Métricas guardadas correctamente.")

Métricas guardadas correctamente.
